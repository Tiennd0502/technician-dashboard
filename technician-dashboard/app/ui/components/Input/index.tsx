import { forwardRef, memo } from 'react';
import {
  FormControl,
  FormLabel,
  Input as ChakraInput,
  InputProps as ChakraInputProps,
  FormErrorMessage,
} from '@chakra-ui/react';

interface InputProps extends ChakraInputProps {
  error?: string;
  label?: string;
}

const Input = forwardRef<HTMLInputElement, InputProps>(function Component(
  { type = 'text', error, label, ...props },
  ref,
) {
  return (
    <FormControl isInvalid={!!error}>
      <FormLabel fontWeight='semiBold'>{label}</FormLabel>
      <ChakraInput
        ref={ref}
        type={type}
        isInvalid={!!error}
        transition='all .5s ease'
        _focusVisible={{
          outline: 'none',
        }}
        {...props}
      />

      {error && <FormErrorMessage>{error}</FormErrorMessage>}
    </FormControl>
  );
});

export default memo(Input);
