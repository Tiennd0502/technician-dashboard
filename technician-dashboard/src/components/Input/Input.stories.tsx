import type { Meta, StoryObj } from '@storybook/react';

import Input from '.';

const meta: Meta<typeof Input> = {
  title: 'Components/Input',
  component: Input,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof Input>;

export const Default: Story = {
  args: {
    label: 'Name:',
  },
};

export const Error: Story = {
  args: {
    error: 'Error message',
    label: 'Name:',
  },
};
