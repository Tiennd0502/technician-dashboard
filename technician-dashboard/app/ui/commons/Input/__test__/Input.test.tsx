import { fireEvent, render } from '@testing-library/react';
import Input from '..';

describe('Input component', () => {
  const onChange = jest.fn();

  test('should show error message when reive error', () => {
    const props = {
      error: 'Error message',
    };

    const { getByText } = render(<Input {...props} />);
    const error = getByText('Error message');

    expect(error).toBeTruthy();
  });

  test('should show label', () => {
    const props = {
      label: 'Name',
    };
    const { getByText } = render(<Input {...props} />);
    const target = getByText('Name');

    expect(target).toBeTruthy();
  });

  it('should call onChange handler when value changes', async () => {
    const props = {
      placeholder: 'Name',
      value: '',
      onChange,
    };

    const { getByPlaceholderText } = render(<Input {...props} />);
    const input = getByPlaceholderText('Name');
    fireEvent.change(input, { target: { value: 'Product' } });

    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('match snapshot', () => {
    const props = {
      error: 'Error message',
      label: 'Name',
    };
    const { container } = render(<Input {...props} />);

    expect(container).toMatchSnapshot();
  });
});
