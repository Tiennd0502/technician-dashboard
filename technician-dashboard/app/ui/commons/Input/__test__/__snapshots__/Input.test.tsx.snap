// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component match snapshot 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    data-invalid=""
    role="group"
  >
    <label
      class="chakra-form__label css-gu124"
      data-invalid=""
      for="field-:r3:"
      id="field-:r3:-label"
    >
      Name
    </label>
    <input
      aria-describedby="field-:r3:-feedback"
      aria-invalid="true"
      class="chakra-input css-1cu97os"
      id="field-:r3:"
      type="text"
    />
    <div
      aria-live="polite"
      class="chakra-form__error-message css-70qvj9"
      id="field-:r3:-feedback"
    >
      Error message
    </div>
  </div>
</div>
`;
